skops/test_persist.py

```diff
@@ -0,0 +1,378 @@
+import tempfile
+from functools import cache
+from pathlib import Path
+
+import numpy as np
+import pytest
+from scipy import special
+from sklearn.base import BaseEstimator
+from sklearn.datasets import load_breast_cancer, load_iris
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import make_scorer
+from sklearn.model_selection import GridSearchCV
+from sklearn.neural_network import MLPRegressor
+from sklearn.pipeline import FeatureUnion, Pipeline
+from sklearn.preprocessing import (
+    FunctionTransformer,
+    MinMaxScaler,
+    PolynomialFeatures,
+    StandardScaler,
+)
+
+from skops._persist import load, save
+
+
+@cache
+def clf_data():
+    X, y = load_iris(return_X_y=True, as_frame=False)
+    return X, y
+
+
+@cache
+def regr_data():
+    X, y = load_breast_cancer(return_X_y=True, as_frame=False)
+    return X, y
+
+
+class TestGeneral:
+    @pytest.fixture(scope="class")
+    def temp_path(self):
+        with tempfile.TemporaryDirectory(prefix="skops-test-temp-path") as temp_path:
+            yield Path(temp_path)
+
+    @pytest.mark.xfail(strict=True)
+    def test_passing_file_handler(self, temp_path):
+        filename = temp_path / "a-file.zip"
+        with open(filename, "w") as f:
+            save("hello world", f)
+        with open(filename, "r") as f:
+            loaded = load(f)
+        assert loaded == "hello world"
+
+    def test_unfitted_estimator(self, temp_path):
+        # also works with unfitted estimators
+        X, y = clf_data()
+        estimator = LogisticRegression(random_state=0, solver="liblinear")
+        filename = temp_path / "model.zip"
+        save(estimator, filename)
+        loaded = load(filename)
+
+        assert loaded.get_params(deep=True) == estimator.get_params(deep=True)
+        loaded.fit(X, y)  # does not raise
+
+    @pytest.mark.xfail(strict=True)
+    def test_adding_custom_methods(self):
+        # yet to be implmeneted
+        assert False
+
+
+def assert_params_equal(params0, params1):
+    # helper function to compare estimator params
+    assert len(params0) == len(params1)
+    assert set(params0.keys()) == set(params1.keys())
+    for key in params0:
+        if key.endswith("steps") or key.endswith("transformer_list"):
+            # TODO: anything smarter?
+            continue
+
+        val0, val1 = params0[key], params1[key]
+        assert type(val0) == type(val1)
+        if isinstance(val0, BaseEstimator):
+            assert_params_equal(val0.get_params(), val1.get_params())
+        elif isinstance(val0, (np.ndarray, np.generic)):
+            np.testing.assert_allclose(val0, val1)
+        else:
+            assert val0 == val1
+
+
+class EstimatorBaseTester:
```

Comment:
For common tests, I rather do the way I now put in the PR: https://github.com/skops-dev/skops/pull/128/files#diff-9ce8f30521c2b9fb6a1b2f0813e522810362fa16cc9a8be1660c4f5637dbcce1

It reduces the boilerplate code I think.